In [1]:
# Import libraries
import pandas as pd
import numpy as np

from surprise import Dataset
from surprise import Reader
from surprise import NormalPredictor
from surprise import SVD
from surprise import KNNWithZScore
from surprise import SVDpp
from surprise import accuracy
from surprise.model_selection import cross_validate
from surprise.model_selection import PredefinedKFold
from surprise.model_selection import train_test_split

from utilities import data_basic_utility as databasic

## File Details

Basic run with SVD without cross validation. This can be the basis for parameter tuning and other stuff later


# Assignment 3 -- Recommendation Systems

* The final challenge is much like Assignment 2 -- but scoped appropriately for the time and your current abilities.
* It is ratings prediction, just like the movielens recommendations we have seen and many other similar problems.
* The features created are based on Beer Reviews from experts on a website.
* Each beer has been scored between 0 and 5 (on a real scale, so 2.75 or 3.5 is OK).
* The official measure is Mean Average Error (MAE) which is pretty intuitive to work with. Everything supports is and it is easy to interpret.
* A set of features have been created based on the reviewer, the written review, and information about the Beer being reviewed.
* Not all features have to be used, and you can easily create new features using the data if you like.
* The features included are:

![title](Images/A3Features.png)

* Sizes of the files are:
|Size | File|
|---|---|
| 1.9G | features.tsv |
| 88B  | header-features.tsv|
| 48B  | header.tsv |
| 15M  | test.tsv |
| 50M  | train.tsv |
| 16M |  val.tsv |


In [2]:
filePrefix = "A3_070_surprise_svd1"
baseDataDir = "C:/Development/Data/COSC2670/Assignment3/A3data/"
subrunDir = "subruns/"
writeSubRunFile = True
seed = databasic.get_random_seed()

In [3]:
# RowID  BeerID  ReviewerID  BeerName  BeerType  Label
#df_train = pd.read_csv(baseDataDir + 'train_wk12.tsv',sep='\t',
df_train = pd.read_csv(baseDataDir + 'train_200k.tsv',sep='\t',
                         names=['RowID','BeerID','ReviewerID',
                                  'BeerName','BeerType','rating'])


#df_vali = pd.read_csv(baseDataDir + 'val_wk12.tsv',sep='\t',
df_vali = pd.read_csv(baseDataDir + 'vali_200k.tsv',sep='\t',
                         names=['RowID','BeerID','ReviewerID',
                                  'BeerName','BeerType','rating'])
df_vali.head(10)


RowID  BeerID  ReviewerID          BeerName                 BeerType  \
0     22   12300        2634     Rauch �r Bock                Rauchbier   
1     27   12300        5634     Rauch �r Bock                Rauchbier   
2     28   12300        3544     Rauch �r Bock                Rauchbier   
3     40   12300        6521     Rauch �r Bock                Rauchbier   
4     43   12300       10177     Rauch �r Bock                Rauchbier   
5     48   12300        2907     Rauch �r Bock                Rauchbier   
6     49   12300        1532     Rauch �r Bock                Rauchbier   
7     50   12300        3452     Rauch �r Bock                Rauchbier   
8     59   12300        6861     Rauch �r Bock                Rauchbier   
9     64    6699        6401  Caldera Pale Ale  American Pale Ale (APA)   

   rating  
0     4.0  
1     4.5  
2     4.5  
3     4.0  
4     4.5  
5     3.5  
6     4.0  
7     3.5  
8     4.0  
9     4.5

In [4]:
df_train.head(10)

RowID  BeerID  ReviewerID       BeerName   BeerType  rating
0     19   12300       10635  Rauch �r Bock  Rauchbier     4.0
1     21   12300        6547  Rauch �r Bock  Rauchbier     4.5
2     23   12300        9789  Rauch �r Bock  Rauchbier     4.5
3     24   12300        7372  Rauch �r Bock  Rauchbier     5.0
4     25   12300        1302  Rauch �r Bock  Rauchbier     4.5
5     26   12300         704  Rauch �r Bock  Rauchbier     4.5
6     29   12300        1747  Rauch �r Bock  Rauchbier     5.0
7     31   12300        9368  Rauch �r Bock  Rauchbier     4.5
8     32   12300        2568  Rauch �r Bock  Rauchbier     4.0
9     33   12300        6838  Rauch �r Bock  Rauchbier     4.0

Column List: 
RowID BrewerID ABV DayofWeek Month DayofMonth Year TimeOfDay Gender Birthday Text Lemmatized POS_Tag

In [5]:
# df_features = pd.read_csv(baseDataDir + 'features-top500.tsv',sep='\t', names=['RowID','BrewerID','ABV','DayofWeek','Month',
df_features = pd.read_csv(baseDataDir + 'features_200k.tsv',sep='\t', names=['RowID','BrewerID','ABV','DayofWeek','Month',
                                                                 'DayofMonth','Year','TimeOfDay','Gender',
                                                                 'Birthday','Text','Lemmatized','POS_Tag'])

df_features.head(10)

RowID  BrewerID  ABV DayofWeek Month  DayofMonth  Year TimeOfDay   Gender  \
0     18      1075  7.4       Mon   Jan           2  2012  15:20:04     Male   
1     19      1075  7.4       Sun   Jan           1  2012  06:46:52     Male   
2     20      1075  7.4       Tue   Nov          29  2011  05:51:44     Male   
3     21      1075  7.4       Sat   Nov           5  2011  22:59:57     Male   
4     22      1075  7.4       Tue   Nov           1  2011  20:40:21     Male   
5     23      1075  7.4       Sat   Oct          29  2011  11:18:34  unknown   
6     24      1075  7.4       Mon   Oct          17  2011  09:04:02  unknown   
7     25      1075  7.4       Tue   Oct          11  2011  10:41:41  unknown   
8     26      1075  7.4       Tue   Oct          11  2011  10:31:22     Male   
9     27      1075  7.4       Mon   Sep          19  2011  14:18:29     Male   

       Birthday                                               Text  \
0       unknown  Pours a murky light brown with a 1 inch fizzy ...   
1       unknown  Faint sudsy head with some with some dissipati...   
2       unknown  A new arrival to the West TN area ... Pours qu...   
3       unknown  Sampled 10/30/11 - Transferring the notes . A ...   
4  Oct 14, 1983  This is my first rauchbier . Pours a burnt amb...   
5       unknown                                                 A    
6       unknown  Pours a mahogany color , rich , with a tan hea...   
7       unknown  Pours light caramel brown with reddish highlig...   
8  Nov 24, 1974  Poured a slightly cloudy deep amber/red color ...   
9  Oct 10, 1988  Big thanks to N2168 for knocking this off my w...   

                                          Lemmatized  \
0  pour a murky light brown with a 1 inch fizzy t...   
1  faint sudsy head with some with some dissipate...   
2  a new arrival to the West TN area ... pour qui...   
3  sample 10/30/11 - transfer the note . a ruby p...   
4  this be my first rauchbier . pour a burn amber...   
5                                                 a    
6  pour a mahogany color , rich , with a tan head...   
7  pour light caramel brown with reddish highligh...   
8  pour a slightly cloudy deep amber/red color wi...   
9  big thanks to n2168 for knock this off my want...   

                                             POS_Tag  
0  VBZ DT JJ NN JJ IN DT CD NN JJ NN NN WDT VBZ I...  
1  NN JJ NN IN DT IN DT VBG JJ NN . JJ JJ NN . DT...  
2  DT JJ NN IN DT NNP NNP NN , VBZ PDT DT NN JJR ...  
3  VBN CD HYPH VBG DT NNS . DT NN VBP IN DT NN NN...  
4  DT VBZ PRP$ JJ NN . VBZ DT VBN NN . JJ NN . NN...  
5                                                DT   
6  VBZ DT NN NN , JJ , IN DT JJ NN . DT NN , VBD ...  
7  NNS JJ NN JJ IN JJ NNS . DT JJ JJ NN VBZ RB VB...  
8  VBD DT RB JJ JJ VBN NN IN DT JJ NN NN IN NN . ...  
9  JJ NNS IN NN IN VBG DT RP PRP$ NNS . VBN IN DT...

In [6]:
idCols = ['RowID','BeerID','ReviewerID']

# Setup the data to be just the Reviewer and the Beer(Item) and the Rating Label we want to learn.
dfTrainFeatures = df_train.drop(['RowID','BeerName','BeerType'],axis=1)
dfValiIds = df_vali[idCols]
dfValiFeatures = df_vali.drop(['RowID','BeerName','BeerType'],axis=1)

dfTrainFeatures.head()

BeerID  ReviewerID  rating
0   12300       10635     4.0
1   12300        6547     4.5
2   12300        9789     4.5
3   12300        7372     5.0
4   12300        1302     4.5

In [7]:

reader = Reader(rating_scale=(0, 5))
dsetTrainFeatures = Dataset.load_from_df(dfTrainFeatures[['BeerID','ReviewerID',
                                    'rating']],reader)

dsetValiFeatures = Dataset.load_from_df(dfValiFeatures[['BeerID','ReviewerID',
                                     'rating']],reader)
trainsetTrainFeatures = dsetTrainFeatures.build_full_trainset()

print(type(dsetTrainFeatures))
print(type(trainsetTrainFeatures))
trainsetTrainFeatures

<class 'surprise.dataset.DatasetAutoFolds'>
<class 'surprise.trainset.Trainset'>


In [8]:

NA,valset = train_test_split(dsetValiFeatures, test_size=1.0)

# best_params (quick) = {'n_epochs': 10, 'lr_all': 0.01, 'reg_all': 0.4}
# best_params (full) = {'n_factors': 30, 'n_epochs': 20, 'biased': True (default), 'lr_all': 0.005, 'reg_all': 0.1}

# SVD with first sets of best params
# algorithm = SVD(n_epochs=10, lr_all=0.01, reg_all=0.4)
algorithm = SVD(n_factors = 30, n_epochs=20, lr_all=0.005, reg_all=0.1)
# algorithm = KNNWithZScore()

model = algorithm.fit(trainsetTrainFeatures)
predictions = algorithm.test(valset)


# Score our predictions with MAE
# It is around 0.77, which means the a random guess based on the distribution of the data
# is on average within 0.77 (plus or minus) the true label.
# Not bad! You can beat it though, I'm sure :).
# Smaller MAE is the better. Good luck!
mae = accuracy.mae(predictions,verbose=True)

print("Average MAE: " + str(mae))

Computing the msd similarity matrix...
Done computing similarity matrix.
MAE:  0.4381
Average MAE: 0.43811115069361556


In [9]:
print(type(valset))
print(valset[0:10])


<class 'list'>
[(1198, 5457, 4.0), (8492, 458, 4.0), (2606, 8562, 4.0), (2160, 8087, 4.0), (10838, 10178, 5.0), (12540, 1387, 4.0), (1623, 7240, 4.5), (9088, 8945, 5.0), (7472, 4779, 4.0), (9640, 6419, 4.5)]


In [10]:
print(type(predictions))
print(str(len(predictions)))
print(predictions[0:10])
print(predictions[0])

<class 'list'>
39509
[Prediction(uid=1198, iid=5457, r_ui=4.0, est=4.322125526106366, details={'actual_k': 40, 'was_impossible': False}), Prediction(uid=8492, iid=458, r_ui=4.0, est=3.8153858607339166, details={'actual_k': 40, 'was_impossible': False}), Prediction(uid=2606, iid=8562, r_ui=4.0, est=3.7394060412177685, details={'actual_k': 12, 'was_impossible': False}), Prediction(uid=2160, iid=8087, r_ui=4.0, est=3.9058132401788654, details={'actual_k': 1, 'was_impossible': False}), Prediction(uid=10838, iid=10178, r_ui=5.0, est=3.759716245324594, details={'actual_k': 40, 'was_impossible': False}), Prediction(uid=12540, iid=1387, r_ui=4.0, est=4.095602534713573, details={'actual_k': 40, 'was_impossible': False}), Prediction(uid=1623, iid=7240, r_ui=4.5, est=4.556510203255987, details={'actual_k': 9, 'was_impossible': False}), Prediction(uid=9088, iid=8945, r_ui=5.0, est=1.8717575154994965, details={'actual_k': 1, 'was_impossible': False}), Prediction(uid=7472, iid=4779, r_ui=4.0, est=3.

In [11]:
# Convert the Predictions to a dataframe so we can lookup predictions easy
lstUIds = list(map(lambda x: x.uid, predictions))
lstIIds = list(map(lambda x: x.iid, predictions))
lstTrueRatings = list(map(lambda x: x.r_ui, predictions))
lstRatingEst = list(map(lambda x: x.est, predictions))


# uid == BeerId, iid == ReviewerId, r_ui == Original Ration, est = Predicted rating
dfPredictions = pd.DataFrame({ "uid": lstUIds,"iid": lstIIds, "r_ui": lstTrueRatings, "Predict": lstRatingEst })

dfPredictions.head()
# dfPredictions[dfPredictions.uid == 3519]

uid    iid  r_ui   Predict
0   1198   5457   4.0  4.322126
1   8492    458   4.0  3.815386
2   2606   8562   4.0  3.739406
3   2160   8087   4.0  3.905813
4  10838  10178   5.0  3.759716

In [12]:
print(dfValiIds.shape)
print(dfPredictions.shape)

(39509, 3)
(39509, 4)


In [13]:
# join the predictions to the ids, sort by rowid and write to file
dfPredictions = pd.merge(dfValiIds, dfPredictions, how="inner", left_on=["BeerID", "ReviewerID"], right_on=["uid", "iid"])
dfPredictions.head()


RowID  BeerID  ReviewerID    uid    iid  r_ui   Predict
0     22   12300        2634  12300   2634   4.0  4.412401
1     27   12300        5634  12300   5634   4.5  4.237398
2     28   12300        3544  12300   3544   4.5  4.460317
3     40   12300        6521  12300   6521   4.0  4.254228
4     43   12300       10177  12300  10177   4.5  4.094124

This basic normal run doesn't give good MAE, so not worth writing out and considering

In [14]:
if writeSubRunFile:
  dfPredictions.sort_values("RowID")[["RowID", "BeerID", "ReviewerID", "Predict"]].to_csv(subrunDir + filePrefix + "_subrun.csv", index=False)

print("Average MAE: " + str(mae))
print("analyse_maes.append(" + str(mae) + ")")
print(dfPredictions.shape)
dfPredictions.sort_values("RowID").head(8)


Average MAE: 0.43811115069361556
analyse_maes.append(0.43811115069361556)
(39703, 7)


RowID  BeerID  ReviewerID    uid    iid  r_ui   Predict
0     22   12300        2634  12300   2634   4.0  4.412401
1     27   12300        5634  12300   5634   4.5  4.237398
2     28   12300        3544  12300   3544   4.5  4.460317
3     40   12300        6521  12300   6521   4.0  4.254228
4     43   12300       10177  12300  10177   4.5  4.094124
5     48   12300        2907  12300   2907   3.5  4.020814
6     49   12300        1532  12300   1532   4.0  4.314729
7     50   12300        3452  12300   3452   3.5  4.119294

In [15]:

print("Run - " + filePrefix)
# Log of Results
analyse_maes = []
analyse_maes.append(0.43352272580226164)
analyse_maes.append(0.43381746394180665)
analyse_maes.append(0.4345411180793805)
analyse_maes.append(0.4345411180793805)

print("Average MAE over all tests: " + str(np.mean(analyse_maes)))

# Make sure it's predicting floats
# dfPredicted["Predict"].drop_duplicates()

Run - A3_070_surprise_svd1
Average MAE over all tests: 0.43410560647570734



Average MAE with simple SVD: 0.43410560647570734

MAE with first run of best params:  0.4311946847059103

MAE with full best params: 0.4280525719411112

MAE with simple SVDpp: 0.429803678058171
